src/svc_xprt.c

```diff
@@ -197,8 +198,26 @@ svc_xprt_lookup(int fd, svc_xprt_setup_t setup)
 	rec = opr_containerof(nv, struct rpc_dplx_rec, fd_node);
 	xprt = &rec->xprt;
 
-	/* lookup reference before unlock ensures shutdown cannot release */
-	SVC_REF(xprt, SVC_REF_FLAG_NONE);
+	/* lookup reference before unlock ensures shutdown cannot release
+	 *
+	 * The current implementation makes it possible to have 0
+	 * refcount xprts in the tree. The code would be easier if
+	 * destroy removes from the tree before decrementing the
+	 * refcount avoiding this possibility. It needs some code
+	 * re-org but easier to implement in the long run!
+	 *
+	 * For now, manually increment the refcnt and see if it indeed
+	 * was zero (1 now!). If so, bailout under tree lock as the xprt
+	 * might be freed as soon as we unlock!
+	 *
+	 * TODO: Need LTTng trace messages for these manual refcounts
+	 */
+	refcount = atomic_inc_int32_t(&xprt->xp_refcnt);
+	if (refcount == 1) { /* proces of getting destroyed */
+		atomic_dec_int32_t(&xprt->xp_refcnt);
+		rwlock_unlock(&t->lock);
+		return NULL;
+	}
```

Comment:
So, I believe this is what the SVC_XPRT_FLAG_DESTROYED check below is for.  Can you explain why this isn't sufficient?  I'm not seeing why that won't work.  Maybe the flag check just needs to be moved inside the lock?

Reply:
Yes, it has to be inside the lock otherwise there is no guaranty of its existence outside the lock. The flag SVC_XPRT_FLAG_DESTROYED is combination of SVC_XPRT_FLAG_DESTROYING and SVC_XPRT_FLAG_RELEASING (I don't think we need to use the combination of two flags, makes it hard to verify the correctness).

Anyway here is my problem:
You call SVC_REF() followed by the SVC_XPRT_FLAG_DESTROYING check:
if the flag is set, how do you undo the SVC_REF? If you call SVC_RELEASE() inside the lock, then this could cause a hang if this thread needs to delete from the hash. Outside the lock is dangerous as it could be freed. So I don't see how flags work correctly other than obfuscating the code.

I am open to fixing it with flags, but I think they make it harder to detect bugs in the code review.

Reply:
I think you do nothing in that case.  It's already in the process of being released/freed, and will be unconditionally freed, so just ignore the extra ref and let it be freed.

Reply:
svc_destroy_it() sets the flag and then call svc_release_it() to decrement the refcount. It is possible that the xprt will never be freed because of this positive refcount.